src/openvic-simulation/economy/production/ResourceGatheringOperation.cpp

```diff
@@ -1,21 +1,353 @@
 #include "ResourceGatheringOperation.hpp"
 
+#include <vector>
+
+#include "openvic-simulation/economy/production/Employee.hpp"
+#include "openvic-simulation/economy/production/ProductionType.hpp"
+#include "openvic-simulation/pop/Pop.hpp"
+#include "openvic-simulation/map/ProvinceInstance.hpp"
+#include "openvic-simulation/map/State.hpp"
+#include "openvic-simulation/modifier/ModifierEffectCache.hpp"
+#include "openvic-simulation/types/fixed_point/FixedPoint.hpp"
+#include "openvic-simulation/utility/Logger.hpp"
+
 using namespace OpenVic;
 
 ResourceGatheringOperation::ResourceGatheringOperation(
-	ProductionType const& new_production_type,
+	ProductionType const* new_production_type_nullable,
 	fixed_point_t new_size_multiplier,
 	fixed_point_t new_revenue_yesterday,
 	fixed_point_t new_output_quantity_yesterday,
 	fixed_point_t new_unsold_quantity_yesterday,
-	ordered_map<Pop*, Pop::pop_size_t>&& new_employees
-) : production_type { new_production_type },
+	std::vector<Employee>&& new_employees,
+	decltype(employee_count_per_type_cache)::keys_t const& pop_type_keys
+) : production_type_nullable { new_production_type_nullable },
 	revenue_yesterday { new_revenue_yesterday },
 	output_quantity_yesterday { new_output_quantity_yesterday },
 	unsold_quantity_yesterday { new_unsold_quantity_yesterday },
 	size_multiplier { new_size_multiplier },
-	employees { std::move(new_employees) } {}
+	employees { std::move(new_employees) },
+	max_employee_count_cache {},
+	total_employees_count_cache {},
+	total_paid_employees_count_cache { 0},
+	total_owner_income_cache {},
+	total_employee_income_cache {},
+	employee_count_per_type_cache { &pop_type_keys }
+{
+	set_production_type_nullable(production_type_nullable);
+}
 
-ResourceGatheringOperation::ResourceGatheringOperation(
-	ProductionType const& new_production_type, fixed_point_t new_size_multiplier
-) : ResourceGatheringOperation { new_production_type, new_size_multiplier, 0, 0, 0, {} } {}
+ResourceGatheringOperation::ResourceGatheringOperation(decltype(employee_count_per_type_cache)::keys_t const& pop_type_keys) : ResourceGatheringOperation {
+	nullptr, fixed_point_t::_0(),
+	 fixed_point_t::_0(), fixed_point_t::_0()
+	 , fixed_point_t::_0(), {}, pop_type_keys
+} {}
+
+//The location ensures only valid RGO production types are passed.
+void ResourceGatheringOperation::set_production_type_nullable(ProductionType const* new_production_type_nullable){
+	production_type_nullable = new_production_type_nullable;
+}
```

Comment:
Why not just make the definition of `production_type_nullable` use `PROPERTY_RW` (which generates a setter identical to this except it's `constexpr` so the function call can be optimised away during compilation) and get rid of this + its header declaration?

Reply:
You're right. I thought I had to recalculate the rgo size multiplier or something but PROPERTY_RW is fine.